### PR TITLE
Blockstore download to have internal loops, this is to ask batches of …

### PR DESCRIPTION
**Pull request by dangershony**
…work to the puller

### CODE REVIEW

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs [hunk 1]
```diff
@@ -35,16 +38,23 @@ public BlockStoreInnerStepFindBlocks(ILoggerFactory loggerFactory)
         {
             this.logger.LogTrace("()");
 
-            while (await ShouldStopFindingBlocks(context) == false)
+            var batchSize = BlockStoreInnerStepContext.DownloadStackThreshold - context.DownloadStack.Count;
+            var batchList = new List<ChainedBlock>(batchSize);
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs]: I don't understand why we can't work with the DownloadStack directly?

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs [hunk 2]
```diff
@@ -35,16 +38,23 @@ public BlockStoreInnerStepFindBlocks(ILoggerFactory loggerFactory)
         {
             this.logger.LogTrace("()");
 
-            while (await ShouldStopFindingBlocks(context) == false)
+            var batchSize = BlockStoreInnerStepContext.DownloadStackThreshold - context.DownloadStack.Count;
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs]: In ShouldStopFindingBlocks we check to see if the DownloadStack's count has exceeded the DownloadStackThreshold so why do we need a batchSize to check? I think it comes to the same thing?

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs [hunk 3]
```diff
@@ -35,16 +38,23 @@ public BlockStoreInnerStepFindBlocks(ILoggerFactory loggerFactory)
...
+					context.StopFindingBlocks();
+	                break;
+	            }
+
+                batchList.Add(context.NextChainedBlock);
                 context.DownloadStack.Enqueue(context.NextChainedBlock);
                 context.GetNextBlock();
             }
 
-            context.StopFindingBlocks();
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs]: Should'nt we always STOP to ask for blocks until a stop condition has been found or the download stack threshold reached?

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepReadBlocks.cs [hunk 4]
```diff
@@ -41,65 +42,69 @@ public BlockStoreInnerStepReadBlocks(ILoggerFactory loggerFactory)
...
             }
+            
 
-            BlockPuller.DownloadedBlock downloadedBlock;
-
-            ChainedBlock nextBlock = context.DownloadStack.Peek();
+            var loopCount = BlockStoreInnerStepContext.DownloadStackThreshold;
 
-            if (context.BlockStoreLoop.BlockPuller.TryGetBlock(nextBlock, out downloadedBlock))
+            while (--loopCount > 0)
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepReadBlocks.cs]: I'm not sure that this is the right thing to do as by the time we get here the outer loop would only be running this InnerStep and as such the loop seems a bit unnecessary? Basically, we'll continue to call TryGetBlock until the DownloadStack is empty, then we'll stop and the outer BS loop will start again.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs (hunk 2 above)
Comment (dangershony) [reply Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs]: In your implementation the code will ask for x blocks and then move on to the read blocks and wait for it to be drained.

That introduces a bottle neck, ideally we want to continuously ask for blocks and store blocks.

Currently the code loops over the two steps together so the result was
Loop over:
- Ask one block
- Take one block (and add to a pending commit list if the list is above a threshold then push)
> this gives a continuation loop that asks and pushes

After you previous change
- ask for 10 blocks
Loop:
- Take one block (and add to a pending commit list if the list is above a threshold then push)

The issue is we wait for the 10 blocks to download before we ask for the next 10 blocks.
The desire is:
- Ask 10 blocks
- 5 blocks arrived 
- Ask for another 5 blocks

**Comment by dangershony** [comment]
I am happy to close this or let you continue in this branch.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs (hunk 2 above)
Comment (fassadlr) [reply Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs]: OK thanks for clarifying @dangershony! I understand now what the issue was with my implementation.

**Comment by dangershony** [comment]
@fassadlr I made some more changes please take over this branch and finish it up.

**Comment by dangershony** [comment]
This is continued on another PR #369

**Pull request closed without merging**